### PR TITLE
[dagster-aws] add ECS Task executor

### LINKED ISSUE #9671: ECS Task Executor
### What's the use case?

Running each op of a Dagster job in its own ECS task (similar to the k8s_job_executor) to make it easier to horizontally scale jobs with many ops, and give better isolation between ops.

### Ideas of implementation

_No response_

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by danielgafni**
## Summary & Motivation

This PR ads a new ECS Tasks Executor.

Currently a `ECSRunLauncher` is required in order to use the new `ecs_executor` and most of the ECS config comes from the run launcher. The executor can override a few important fields (such as resources and container overrides). 

Resolve #9671

## How I Tested These Changes

- manual testing in personal AWS account
- adding tests with mocks. Heavily using the existing run launcher testing infra since the executor currently requires it anyway

## Changelog

[dagster-aws] new `ecs_executor` which executes Dagster steps via AWS ECS tasks. Initially, we expect it to be used in conjunction with `ECSRunLauncher`.

### CODE REVIEW

**Comment by danielgafni** [comment]
* **#26098** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26098?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26098?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by danielgafni** [comment]
Hey @gibsondan @alangenfeld, asking for an initial review here. 

I have tested the ECS Executor in my personal AWS account. Adding some tests involving mocks too.

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+            " always be set by the run launcher."
+        ),
+    ),
+    "cpu": Field(IntSource, is_required=False),
+    "memory": Field(IntSource, is_required=False),
+    "ephemeral_storage": Field(IntSource, is_required=False),
+    "task_overrides": Field(
+        Permissive({}),
+        is_required=False,
+    ),
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:58]: recognizing that this was all on my original PR - there's some overlap between these parameters right? All of them can be set on run_task_kwargs. I think it makes sense to put things like cpu/memory as top-level fields since they are fields that most people will want to change

What happens if you set 'overrides' on the run_task_kwargs param though? should we not allow that? should we merge them? how do we decide which one wins if you set both "task_overrides.cpu" and "cpu"? What if you set task_overrides.containerOverrides?

the snake case vs. camel case thing is kind of a bummer too (dagster is generally snake case but the boto apis expect camelCase, so you end up with grossness like task_overrides.containerOverrides), but I don't see an obvious way around that

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    def _get_run_task_kwargs(
+        self,
+        run: DagsterRun,
+        args: Sequence[str],
+        step_key: str,
+        step_tags: Mapping[str, str],
+        step_handler_context: StepHandlerContext,
+        container_context: EcsContainerContext,
+    ):
+        run_launcher = cast(EcsRunLauncher, step_handler_context.instance.run_launcher)
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:156]: lets use more things from the run launcher here when possible, see above - if you have specified what cluster to use in your dagster.yaml for the run launcher we should be able to pull that here has well rather than re-deriving it from the current task

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+        current_task_metadata = get_current_ecs_task_metadata()
+        current_task = get_current_ecs_task(
+            self.ecs, current_task_metadata.task_arn, current_task_metadata.cluster
+        )
+        task_kwargs = get_task_kwargs_from_current_task(
+            self.ec2,
+            current_task_metadata.cluster,
+            current_task,
+        )
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:138]: hmmmm I think using the same task definition as the run / calling task makes sense - but in many cases (notably dagster+ hybrid and serverless agents both) the run task kwargs are explicitly passed in via the dagster.yaml rather than derived from the current task. Can we delegate to the run launcher to determine these instead so that that logic always matches?

Comment (gibsondan) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:138]: That might require some changes to the run launcher logic to cleanly pull out the bit of logic that you would need - but that might actually be ok? What bad things if any would happen if this logic was changed to

`task_kwargs = self.run_launcher._run_task_kwargs(...)` ? (potentially the protected method would need to be made public)

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+    def _get_container_context(
+        self, step_handler_context: StepHandlerContext
+    ) -> EcsContainerContext:
+        return EcsContainerContext.create_for_run(
+            step_handler_context.dagster_run,
+            cast(EcsRunLauncher, step_handler_context.instance.run_launcher),
+        )
+
+    def _run_task(self, **run_task_kwargs):
+        print(json.dumps(run_task_kwargs))  # noqa
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:240]: will remove before landing i assume?

**Review by gibsondan — CHANGES_REQUESTED** [review]
broadly makes sense to me - i think the major outstanding question is how much to delegate to the run launcher, the prototype I wrote originally always looked at the current ECS task but that was before you discovered that we also needed to re-compute the run_task_kwargs potentially. I think we want to pull those from the dagster.yaml if they are set.

Do we want to make this experimental at first? or no?

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py (hunk 3 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:142]: Thinking about this again... why is using the run launcher better than using the current task? 

1. When Dagster is deployed with the ECS run launcher, the current task will always be created by the Run Launcher, and thus should contain configuration coming from the run launcher anyway. 
2. Not relying on the ECS run launcher will open up the possibility of using the ECS executor with the default run launcher (but when Dagster itself is deployed on ECS)
3. The run launcher config is not available during executor initialization (unlike the current task), because it needs the dagster run in order to call `ECSRunLauncher._run_task_kwargs`. This leads to more convoluted executor logic, because we will have to manage some state to save these kwargs during the `launch_step` method execution.

**Comment by danielgafni** [comment]
@gibsondan hey Daniel, I'm now pulling base configuration from the run launcher.

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py [hunk 5]
```diff
@@ -429,39 +432,16 @@ def _get_run_tags(self, run_id: str) -> Tags:
     def _get_command_args(self, run_args: ExecuteRunArgs, context: LaunchRunContext):
         return run_args.get_command_args()
 
-    def _get_image_for_run(self, context: LaunchRunContext) -> Optional[str]:
-        job_origin = check.not_none(context.job_code_origin)
-        return job_origin.repository_origin.container_image
+    @staticmethod
+    def get_image_for_run(run: DagsterRun) -> Optional[str]:
+        return (
+            run.tags.get("dagster/image") or run.job_code_origin.repository_origin.container_image
```
Comment (danielgafni) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:438]: @gibsondan drive-by: it makes sense to check `dagster/image` for a manual override right? 

I'm also setting it in tests

**Review by gibsondan — APPROVED** [review]
a few important inlines before landing, particularly the part about determining the run launcher kwargs in the constructor rather than within launch_step

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+            run_task_kwargs = {**run_task_kwargs, **json.loads(kwargs_from_tags)}
+
+        run_task_kwargs["tags"] = [
+            *run_task_kwargs.get("tags", []),
+            *[
+                {"key": k, "value": v}
+                for k, v in step_handler_context.dagster_run.dagster_execution_info.items()
+            ],
+            {"key": "dagster/step-key", "value": step_key},
+        ]
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:167]: No it doesn't. Subsequent entries override the previous.

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+                    "command": args,
+                    "environment": executor_env_vars,
+                }
+            )
+
+        if isinstance(run_launcher, EcsRunLauncher):
+            # get ECSRunLauncher run_task kwargs and override them
+            image = run_launcher.get_image_for_run(run)
+            run_task_kwargs = deep_merge_dicts(
+                run_launcher.get_run_task_kwargs(run, image, container_context), run_task_kwargs
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:220]: > I'm not sure off the top of my head if setting this as a tag in the launchpad would have any effect - I suspect it would not? i think it would still get overridden. This would work for checking the tag after the run was launched though

It should work. Both run launcher and executor code will (now) pick up the user-provided image. 
That's how it works with the K8s run launcher / executor.

File: python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/executor_tests/test_executor.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+    return StepHandlerContext(
+        instance=instance,
+        plan_context=plan_context,
+        steps=execution_plan.steps,
+        execute_step_args=execute_step_args,
+    )
+
+
+def test_executor_init(instance_cm: Callable[..., ContextManager[DagsterInstance]]):
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/executor_tests/test_executor.py:90]: I'll add some assertions to this test (it should already be using a ECS run launcher)

[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws_tests/ecs_tests/launcher_tests/executor_tests/test_executor.py:90]: Added a check for `launchType` which is coming from the run launcher

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:220]: I only found one `add_run_tags` usage in the run launcher code and it does not set the `dagster/image` tag.
Looks like this is safe to do. 
Additionally, we should set the `dagster/image` tag when the image comes from the code location config (`DAGSTER_CURRENT_IMAGE` or cli argument)

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+            )
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:221]: fair point, did that

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+        self._cpu = str(cpu) if cpu else None
+        self._memory = str(memory) if memory else None
+
+        self._ephemeral_storage = ephemeral_storage
+        self._task_overrides = check.opt_mapping_param(task_overrides, "task_overrides")
+
+        current_task_metadata = get_current_ecs_task_metadata()
+        current_task = get_current_ecs_task(
+            self.ecs, current_task_metadata.task_arn, current_task_metadata.cluster
+        )
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:132]: Yeah looks like we might need a well-defined hook for state mutations which use the run object like this one

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py (hunk 5 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py]: we decided to remove this logic

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py]: we decided to remove this logic

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+        else:
+            current_task_kwargs = {}
+
+        run_launcher_kwargs = {**current_task_kwargs, **run_launcher.run_task_kwargs}
+
+        self._cluster_arn = current_task["clusterArn"]
+        self._task_definition_arn = current_task["taskDefinitionArn"]
+
+        # note: more kwargs will be pulled from the run launcher
+        # during launch_step
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:155]: rm, no longer true

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+        self._run_task_kwargs = {
+            **run_launcher_kwargs,
+            **run_task_kwargs,
+            "taskDefinition": current_task["taskDefinitionArn"],
+        }
+
+        # TODO: change launch_step to return task ARN
+        # this will be a breaking change so we need to wait fore a minor release
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:163]: nit fore => for

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+
+        self._cluster_arn = current_task["clusterArn"]
+        self._task_definition_arn = current_task["taskDefinitionArn"]
+
+        # note: more kwargs will be pulled from the run launcher
+        # during launch_step
+        self._run_task_kwargs = {
+            **run_launcher_kwargs,
+            **run_task_kwargs,
+            "taskDefinition": current_task["taskDefinitionArn"],
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:159]: if the user has container_context.task_definition_arn set I think we want to use that here instead

I also think this order might not be quite right? This should be first in the list so that you can override it via config if you want

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+        ]
+
+        if run.remote_job_origin:
+            run_task_kwargs["tags"] = [
+                *run_task_kwargs["tags"],
+                {
+                    "key": "dagster/code-location",
+                    "value": run.remote_job_origin.repository_origin.code_location_origin.location_name,
+                },
+            ]
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:205]: i think you want to call run_launcher.build_ecs_tags_for_run_task? so that it also incorporates tags that are set on the EcsContainerContext (for example, if in cloud they have set tags in their dagster_cloud.yaml, those should probably get propagated to the step tasks as well?)

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+        if ephemeral_storage:
+            overrides["ephemeralStorage"] = {"sizeInGiB": int(ephemeral_storage)}
+
+        if tag_overrides := step_tags.get("ecs/task_overrides"):
+            overrides = deep_merge_dicts(overrides, json.loads(tag_overrides))
+
+        return overrides
+
+    def _get_step_id(self, step_handler_context: StepHandlerContext):
+        """Step ID is used to identify the ECS task in the ECS cluster. It is unique to specific step being executed and takes into account op-level retries. It's used as a workardoung to avoid having to return task ARN from launch_step."""
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:273]: nit typo in workardoung

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+
+        step_id = self._get_step_id(step_handler_context)
+
+        self._launched_tasks[step_id] = task["taskArn"]
+
+    def check_step_health(self, step_handler_context: StepHandlerContext) -> CheckStepHealthResult:
+        step_key = self._get_step_key(step_handler_context)
+
+        task_arn = self._launched_tasks.get(self._get_step_id(step_handler_context))
+        cluster_arn = self._cluster_arn
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:350]: probably want some more explicit error handling here if a task with that ID can't be found ( should never happen )

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+
+                return CheckStepHealthResult.unhealthy(reason=cluster_failure_info)
+
+        return CheckStepHealthResult.healthy()
+
+    def terminate_step(
+        self,
+        step_handler_context: StepHandlerContext,
+    ) -> None:
+        task_arn = self._launched_tasks.pop(self._get_step_id(step_handler_context))
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:384]: same here re: error handling

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+        cluster_arn = self._cluster_arn
+        step_key = self._get_step_key(step_handler_context)
+
+        DagsterEvent.engine_event(
+            step_handler_context.get_step_context(step_key),
+            message=f"Deleting task {task_arn} for step",
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:390]: s/Deleting/Stopping

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py [hunk 19]
```diff
@@ -429,39 +432,16 @@ def _get_run_tags(self, run_id: str) -> Tags:
     def _get_command_args(self, run_args: ExecuteRunArgs, context: LaunchRunContext):
         return run_args.get_command_args()
 
-    def _get_image_for_run(self, context: LaunchRunContext) -> Optional[str]:
-        job_origin = check.not_none(context.job_code_origin)
-        return job_origin.repository_origin.container_image
+    @staticmethod
+    def get_image_for_run(run: DagsterRun) -> Optional[str]:
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:436]: still needed?

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py [hunk 20]
```diff
@@ -644,10 +624,10 @@ def _get_current_task(self):
     def _get_run_task_definition_family(self, run: DagsterRun) -> str:
         return get_task_definition_family("run", check.not_none(run.remote_job_origin))
 
-    def _get_container_name(self, container_context: EcsContainerContext) -> str:
+    def get_container_name(self, container_context: EcsContainerContext) -> str:
         return container_context.container_name or self.container_name
 
-    def _run_task_kwargs(
+    def get_run_task_kwargs(
```
Comment (gibsondan) [inline python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:630]: still needed?

**Review by gibsondan — APPROVED** [review]

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py (hunk 19 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py:436]: I'd say yes for slightly better readability

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py (hunk 13 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:159]: > if the user has container_context.task_definition_arn set I think we want to use that here instead

This will come from `run_launcher_kwargs` right?

Comment (gibsondan) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:159]: We don't know the container context yet, it is run-dependant - so it's not actually possible for it to have it here yet. We would need to check it and set it later.

the relevant code from the run launcher is here: https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py#L661-L664

File: python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py (hunk 16 above)
Comment (danielgafni) [reply python_modules/libraries/dagster-aws/dagster_aws/ecs/executor.py:350]: done

**Comment by danielgafni** [comment]
Thank you @gibsondan for your careful review!

**Comment by danielgafni** [comment]
### Merge activity

* **Dec 17, 6:47 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/26098).
